### PR TITLE
Final KB REST API

### PR DESCRIPTION
**Pull request by marcomicera**
## Tasks

- [x] [Endpoints table and their progress](https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects)
- ~~API specifications~~ (dropped in d13d671a, response data structures already generated by JOOQ)
- [x] Server stubs 0c9eea82
- [x] Queries
- [ ] Tests
- [x] Pagination (#88, d1b7acba0581acc0efe261bb4149ddbd8f49ef49)

### CODE REVIEW

**Comment by marcomicera** [comment]
Hey @proksch & @ilyagrishkov, I wrote everything down in this [response codes table](https://github.com/fasten-project/fasten/blob/kbapi/analyzer/restapi-plugin/design/response_codes.md), would you guys mind double-checking it real quick? (I'd advise you to open that in an editor as it's pretty wide and GitHub sort of squeezes it to fit in its narrow layout).

Also, I (once again) dropped the idea of writing OpenAPI specifications. I didn't think of how much time it would have required to transcribe all these data structures into oas3 back then.

Thanks!

**Comment by marcomicera** [comment]
I'm working on the REST server already, will be pushing some stubs soon.
Reporting eventual endpoint changes will be trivial until I start writing some queries (I'd need access to the SQL+graph DBs though!).

**Comment by ilyagrishkov** [comment]
> Hey @proksch & @ilyagrishkov, I wrote everything down in this [response codes table](https://github.com/fasten-project/fasten/blob/kbapi/analyzer/restapi-plugin/design/response_codes.md), would you guys mind double-checking it real quick? (I'd advise you to open that in an editor as it's pretty wide and GitHub sort of squeezes it to fit in its narrow layout).
> 
> Also, I (once again) dropped the idea of writing OpenAPI specifications. I didn't think of how much time it would have required to transcribe all these data structures into oas3 back then.
> 
> Thanks!

There are some updates on the queries. Basically we remove all queries that return information that isn't already in the database. Here is the list of queries that should be removed from the list: 
- `/mvn/packages/{pkg}/{pkg_ver}/vulnerabilities`
- `/mvn/packages/{pkg}/{pkg_ver}/callgraph` (fully merged one)
- `/mvn/packages/{pkg}/{pkg_ver}/deps` (resolved ones)
- `/mvn/packages/{pkg}/{pkg_ver}/{fasten_uri}/vulnerabilities`
- `/mvn/packages/{pkg}/{pkg_ver}/{fasten_uri}/reach`

Also the query that returns all information about package_version (`/mvn/packages/{pkg}/{pkg_ver}`) should include metadata like all other queries that return "all information" from a table.

**Comment by marcomicera** [comment]
Nice, thanks for the update, I'll be reporting those changes immediately.
I'll just be commenting out those lines so that we can easily restore them in the future.

One thing though:

> * `/mvn/packages/{pkg}/{pkg_ver}/callgraph` (fully merged one)

You mean the "resolved" one, right?

**Comment by ilyagrishkov** [comment]
> Nice, thanks for the update, I'll be reporting those changes immediately.
> I'll just be commenting out those lines so that we can easily restore them in the future.
> 
> One thing though:
> 
> > * `/mvn/packages/{pkg}/{pkg_ver}/callgraph` (fully merged one)
> 
> You mean the "resolved" one, right?

Yes, the resolved one

**Comment by marcomicera** [comment]
Cool, done in https://github.com/fasten-project/fasten/pull/80/commits/47ad9ceeb7be6b190df26e7088a4744a73ff96db. Now I only need the KB URL to move forward.

**Comment by gousiosg** [comment]
@marcomicera can you move the REST API specs to the Wiki? Also, indicate which endpoints are done and which are not.

**Comment by marcomicera** [comment]
Sure.

**Comment by marcomicera** [comment]
@fullsushidev in case you wanna join:

1. [Here](https://github.com/fasten-project/fasten/tree/kbapi/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/mvn) I'll be adding API classes and service interfaces.
1. I'll be implementing those services in the [`impl`](https://github.com/fasten-project/fasten/tree/kbapi/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/mvn/impl) subpackage.
1. There's no CDI at the moment, we'll need it for testing.
    I'll be using `@Inject`ion when the time comes:
    https://github.com/fasten-project/fasten/blob/028af6098c7ccfc6591bf1286e8662fc9a963cc5/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/mvn/PackageApi.java#L14-L16
    And auto-discovery:
    https://github.com/fasten-project/fasten/blob/028af6098c7ccfc6591bf1286e8662fc9a963cc5/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/RestApplication.java#L14-L16

**Comment by gousiosg** [comment]
Hi @marcomicera, great progress so far! I am missing tests though. Are you going to add them later?

**Comment by marcomicera** [comment]
> Hi @marcomicera, great progress so far!

Thanks!

> I am missing tests though. Are you going to add them later?

Yes, I thought those endpoints were urgent, so I postponed testing at the end (also paging is missing, you can see all these details in the [endpoints table](https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects) in which I also track my progress).\
If no one is going to use this REST API without me writing tests first (which totally makes sense), I can easily change my priorities for the next two days.

**Comment by marcomicera** [comment]
If you need the API urgently @tmortagne @vmassol, here it is.

Here's the [list of endpoints and their current status](https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects).
As you can see, a couple of endpoints might fail due to Jetty timeouts, and we're still missing unit tests, but it might be enough for internal use.

**Comment by tmortagne** [comment]
OK thanks for the documentation @marcomicera !

Your message seems to suggest it's deployed somewhere but I must have missed what is the base URL.

**Comment by marcomicera** [comment]
> Your message seems to suggest it's deployed somewhere but I must have missed what is the base URL.

It has not been deployed yet, sorry if I haven't been clear enough. For that, I guess I have to ask to @proksch: can we deploy it somewhere, perhaps on the same Monster server I have access to?

---

@tmortagne At the moment, you can only [launch it locally](https://github.com/fasten-project/fasten/blob/kbapi/analyzer/restapi-plugin/README.md#usage):

> ## Usage
> 
> 1. Port-forward the remote Postgres server to a local port, e.g., `5433`:    
>     ```bash
>     export KB_USER=...
>     export KB_ADDR=...
>     ssh -f ${KB_USER}@${KB_ADDR} -L 5433:${KB_ADDR}:5432 -N
>     ```
> 1. Launch the REST server (example from the root directory and with port `9090`):
>     ```bash
>     PGPASSWORD=... mvn clean install exec:java \
>       -f analyzer/restapi-plugin/pom.xml \
>       -Dexec.args='--port 9090 --database jdbc:postgresql://localhost:5433/<DB_NAME> --username <KB_USER>'
>     ```

**Comment by tmortagne** [comment]
OK thanks for the details @marcomicera.

**Comment by proksch** [comment]
> can we deploy it somewhere, perhaps on the same Monster server I have access to?

The easiest option would be if we could run it as a service in kubernetes. I guess to get started and to give everybody access as soon as possible, a temporary "standalone" server would be ok too. Since we have an nginx proxy running in front (on lima), the change under the hood should be transparent anyways...

**Review by lisa-noeth — APPROVED** [review]

**Comment by gousiosg** [comment]
I deployed this on our infrastructure and it seems like it is producing some output, see for example [this URL](http://lima.ewi.tudelft.nl/api/mvn/packages/org.eclipse.jetty:jetty-webapp/versions). Things that need to be fixed (in reference to the URL above):

- [ ] The output seems to be a JSON dump of a JOOQ record. We need to remove the `fields` field and replace `records` with just its contents
- [ ] There is excessive quoting. For example this `"{\"access\": \"protected\", \"defined\": true}"` can be simply replaced by `"{"access": "protected",  "defined": true}`

**Comment by gousiosg** [comment]
Moreover, the query that is used to serve the versions endpoint is wrong 

```sql
 select "public"."package_versions"."id", "public"."package_versions"."package_id", "public"."package_versions"."version", "public"."package_versions"."cg_generator", "public"."package_versions"."created_at", "public"."package_versions"."metadata", "public"."callables"."metadata" as "callable_metadata" from "public"."packages" join "public"."package_versions" on "public"."packages"."id" = "public"."package_versions"."package_id" join "public"."modules" on "public"."package_versions"."id" = "public"."modules"."package_version_id" join "public"."callables" on "public"."modules"."id" = "public"."callables"."module_id" where lower("public"."packages"."package_name") = lower('org.eclipse.jetty:jetty-webapp') limit 10                       
```

Why do you need to join with `callables` to retrieve the versions?

**Comment by marcomicera** [comment]
> Why do you need to join with `callables` to retrieve the versions?

I've had a call with @proksch and @ilyagrishkov during which we agreed on all the fields to be returned by each and every endpoint. I think `callables`' `metadata` were included as a consequence of your endpoint description in your first email:

> `/mvn/packages/junit:junit/versions` <- info about versions of a package, **including potential vulnerabilities**. Shows basic information about the package, processed versions and their metadata, including a range of vulnerable versions.

How should it be changed, i.e., where should I retrieve vulnerabilities from?

**Comment by marcomicera** [comment]
> [edit] just noticed that there is a `limit 10` at the end of the query. Why?

Just a default value needed for pagination.
Can be changed here:

https://github.com/fasten-project/fasten/blob/e2602b67cda88adce5ebdb77b158a81bf28db620/analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/RestApplication.java#L14

**Comment by gousiosg** [comment]
From version `metadata`. Vulnerabilities will be written first there and propagated to the appropriate `callables`

**Comment by marcomicera** [comment]
> From version `metadata`. Vulnerabilities will be written first there and propagated to the appropriate `callables`

Will do!

**Comment by marcomicera** [comment]
> * [ ]  The output seems to be a JSON dump of a JOOQ record. We need to remove the `fields` field and replace `records` with just its contents
> * [ ]  There is excessive quoting. For example this `"{\"access\": \"protected\", \"defined\": true}"` can be simply replaced by `"{"access": "protected",  "defined": true}`

Thanks for noticing, I created #100 out of this with a simple right-click on your comment.

**Comment by gousiosg** [comment]
General design question; why do we have different endpoints for `{entity}` and `{entity}/metadata`, for example `{pkg_ver}` and `{pkg_ver}/metadata`? When retrieving an entity, I would expect all related metadata to be retrieved as well.

**Comment by marcomicera** [comment]
> General design question; why do we have different endpoints for `{entity}` and `{entity}/metadata`, for example `{pkg_ver}` and `{pkg_ver}/metadata`? When retrieving an entity, I would expect all related metadata to be retrieved as well.

I've been following your first endpoints structure you proposed via email some time ago.

For your convenience, here it is:

> All endpoints should accept paging arguments (e.g. per_page=50&page=1)
> 
> Packages and package versions
> /mvn/packages/junit:junit/versions <- info about versions of a package, including potential vulnerabilities. Shows basic information about the package, processed versions and their metadata, including a range of vulnerable versions.
> /mvn/packages/junit:junit/0.4 <- all info about pkg version
> /mvn/packages/junit:junit/0.4/metadata <- all metadata of pkg version
> /mvn/packages/junit:junit/0.4/vulnerabilities <- all info about pkg version vulnerabilities
> /mvn/packages/junit:junit/0.4/callgraph <- pkg call graph
> /mvn/packages/junit:junit/0.4/callgraph?resolve=<timestamp>&full=T <- fully stitched pkg callgraph
> 
> Dependencies
> /mvn/packages/junit:junit/0.4/deps <- (unresolved) deps for pkg version
> /mvn/packages/junit:junit/0.4/deps?resolve=<timestamp> <- resolve deps for timestamp
> 
> Modules
> /mvn/packages/junit:junit/0.4/modules <- all info about modules of pkg + metadata
> /mvn/packages/junit:junit/0.4/modules/<namespace>/metadata <- metadata for a module
> /mvn/packages/junit:junit/0.4/modules/<namespace>/files <- file info for a module
> 
> Binary Modules
> /mvn/packages/junit:junit/0.4/binary-modules <- all info about binary modules of pkg + metadata
> /mvn/packages/junit:junit/0.4/binary-modules/<name>/metadata <- metadata for a binary module
> /mvn/packages/junit:junit/0.4/binary-modules/<name>/files <- file info for a binary module
> 
> Callables
> /mvn/packages/junit:junit/0.4/callables <- callables + metadata
> /mvn/packages/junit:junit/0.4/<fasten_uri>/vulnerabilities <- known vulnerabilities on callable
> /mvn/packages/junit:junit/0.4/<fasten_uri>/metadata <- metadata of the callable
> /mvn/packages/junit:junit/0.4/<fasten_uri>/reach <- packages dependent on a given callable
> 
> Edges
> /mvn/packages/junit:junit/0.4/edges <- edges + metadata
> 
> Files
> /mvn/packages/junit:junit/0.4/files <- files + metadata

I agree with that specific change though, I would personally do the same.

Too bad that we cannot make PRs for Wiki pages, that would be helpful in this case.
What if we move [the endpoints table](https://github.com/fasten-project/fasten/wiki/API-endpoints-for-Maven-projects) in a README file and then keep track of all the changes we wanna bring to the API in a PR? @proksch @gousiosg what do you think?

That way people can propose changes to the API in a more orderly manner IMO.

**Comment by romatallinn** [comment]
@marcomicera it'd be useful to have an endpoint /{pkg} that retrieves the Package instance (forge, repo, etc) with the latest PackageVersion of it.

File: analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/mvn/PackageApi.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @Path("/{pkg}/{pkg_ver}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getPackageVersion(@PathParam("pkg") String package_name,
+                                      @PathParam("pkg_ver") String package_version,
+                                      @DefaultValue("0")
+                                      @QueryParam("offset") short offset,
+                                      @DefaultValue(RestApplication.DEFAULT_PAGE_SIZE)
+                                      @QueryParam("limit") short limit) {
+        return service.getPackageVersion(package_name, package_version, offset, limit);
+    }
```
Comment (romatallinn) [inline analyzer/restapi-plugin/src/main/java/eu/fasten/analyzer/restapiplugin/api/mvn/PackageApi.java:37]: - I believe the response of this endpoint should be flattened. We expect one PackageVersion entity always, but API returns it with a nested array, as getPackageVersions does. While `getPackageVersions` actually may have multiple instances, `getPackageVersion` may not.
- Is there a point in having pagination for such endpoints? I mean, again, we don't expect multiple instances in return, but just one specific?

**Comment by marcomicera** [comment]
> @marcomicera it'd be useful to have an endpoint `/{pkg}` that retrieves the `Package` instance _(forge, repo, etc)_ with the latest `PackageVersion` of it.

Please create an issue for that (and add it to the [REST API project](https://github.com/fasten-project/fasten/projects/10), it's hard for me to keep track of all these requests from all different places.

File: analyzer/restapi-plugin/README.md [hunk 2]
```diff
@@ -45,6 +45,14 @@ The [server configuration file](src/main/resources/application.properties) has t
 
 <!-- TODO ### Requirements  -->
 
+## Postman workspace
+
+You can find a
+[Postman workspace](https://learning.postman.com/docs/getting-started/importing-and-exporting-data)
+configuration
+[here](endpoints.postman_collection.json)
+.
```
Comment (marcomicera) [inline analyzer/restapi-plugin/README.md:54]: @MihhailSokolov you should find everything you need to work with Postman here.

**Comment by gousiosg** [comment]
We should merge this PR at one point. I don't believe there is any outstanding endpoint currently.

**Comment by MihhailSokolov** [comment]
> We should merge this PR at one point. I don't believe there is any outstanding endpoint currently.

Yes, I want to fix issue #197 first, then it should be reviewed (if possible) and merged.

**Comment by MihhailSokolov** [comment]
The #197 is resolved, so the PR is ready for review and merge

**Comment by romatallinn** [comment]
I will work on #205 tomorrow. It would be nice to include it in the merge.

**Review by romatallinn — APPROVED** [review]

**Pull request merged**